File: packages/py-common/src/py_common/schemas.py
"""Common Pydantic schemas for ML model services."""

from pydantic import BaseModel, Field


class ModelPredictRequest(BaseModel):
    """Request schema for model prediction endpoints."""

    text: str = Field(..., description="Text to analyze", max_length=50000)
    request_id: str = Field(..., description="Request ID for tracing")


class ModelPredictResponse(BaseModel):
    """Response schema from model prediction endpoints."""

    flagged: bool = Field(..., description="Whether the text was flagged")
    score: float = Field(..., ge=0.0, le=1.0, description="Confidence score")
    details: list[str] = Field(default_factory=list, description="Explanation details")
    latency_ms: int = Field(..., ge=0, description="Inference latency in milliseconds")


class ValidateRequest(BaseModel):
    """Request schema for the main validation endpoint."""

    request_id: str | None = Field(default=None, description="Optional client-provided request ID")
    project_id: str = Field(..., description="Project ID for config lookup")
    text: str = Field(..., max_length=50000, description="Text to validate")
    type: str = Field(default="input", pattern="^(input|output)$", description="Input or output")
    metadata: dict | None = Field(default=None, description="Optional metadata")


class ModelResultResponse(BaseModel):
    """Result from a single model."""

    flagged: bool
    score: float
    details: list[str]
    latency_ms: int


class ValidateResponse(BaseModel):
    """Response schema for the main validation endpoint."""

    request_id: str = Field(..., description="Request ID")
    flagged: bool = Field(..., description="Overall flag status")
    flag_reasons: list[str] = Field(default_factory=list, description="Reasons for flagging")
    model_results: dict[str, ModelResultResponse] = Field(
        default_factory=dict, description="Per-model results"
    )
    partial_failure: bool = Field(default=False, description="Some models failed")
    failed_models: list[str] = Field(default_factory=list, description="Models that failed")
    latency_ms: int = Field(..., ge=0, description="Total request latency")

File: packages/go-common/schemas.go
// Package gocommon provides shared schemas for the guardrails platform.
package gocommon

// ModelPredictRequest is the request schema for model prediction endpoints.
// Matches the Python ModelPredictRequest in py-common.
type ModelPredictRequest struct {
	// Text to analyze (max 50000 characters)
	Text string `json:"text"`
	// Request ID for tracing
	RequestID string `json:"request_id"`
}

// ModelPredictResponse is the response schema from model prediction endpoints.
// Matches the Python ModelPredictResponse in py-common.
type ModelPredictResponse struct {
	// Whether the text was flagged
	Flagged bool `json:"flagged"`
	// Confidence score (0.0 to 1.0)
	Score float64 `json:"score"`
	// Explanation details
	Details []string `json:"details"`
	// Inference latency in milliseconds
	LatencyMs int `json:"latency_ms"`
}

// ValidateRequest is the request schema for the main validation endpoint.
// Matches the Python ValidateRequest in py-common.
type ValidateRequest struct {
	// Optional client-provided request ID
	RequestID string `json:"request_id,omitempty"`
	// Project ID for config lookup
	ProjectID string `json:"project_id"`
	// Text to validate (max 50000 characters)
	Text string `json:"text"`
	// Input or output type (default: "input")
	Type string `json:"type"`
	// Optional metadata
	Metadata map[string]interface{} `json:"metadata,omitempty"`
}

// ModelResultResponse is the result from a single model.
// Matches the Python ModelResultResponse in py-common.
type ModelResultResponse struct {
	Flagged   bool     `json:"flagged"`
	Score     float64  `json:"score"`
	Details   []string `json:"details"`
	LatencyMs int      `json:"latency_ms"`
}

// ValidateResponse is the response schema for the main validation endpoint.
// Matches the Python ValidateResponse in py-common.
type ValidateResponse struct {
	// Request ID
	RequestID string `json:"request_id"`
	// Overall flag status
	Flagged bool `json:"flagged"`
	// Reasons for flagging
	FlagReasons []string `json:"flag_reasons"`
	// Per-model results
	ModelResults map[string]ModelResultResponse `json:"model_results"`
	// Some models failed
	PartialFailure bool `json:"partial_failure"`
	// Models that failed
	FailedModels []string `json:"failed_models"`
	// Total request latency in milliseconds
	LatencyMs int `json:"latency_ms"`
}

// HealthResponse is the response for health check endpoints.
type HealthResponse struct {
	Status string `json:"status"`
	Model  string `json:"model,omitempty"`
}

// ReadyResponse is the response for readiness check endpoints.
type ReadyResponse struct {
	Status          string   `json:"status"`
	AvailableModels []string `json:"available_models,omitempty"`
	Model           string   `json:"model,omitempty"`
}

// ErrorResponse is the standard error response format.
type ErrorResponse struct {
	Error ErrorDetail `json:"error"`
}

// ErrorDetail contains error information.
type ErrorDetail struct {
	Code    string                 `json:"code"`
	Message string                 `json:"message"`
	Details map[string]interface{} `json:"details,omitempty"`
}

// CircuitBreakerStatus represents the status of a circuit breaker.
type CircuitBreakerStatus struct {
	Name            string  `json:"name"`
	State           string  `json:"state"`
	FailureCount    int     `json:"failure_count"`
	SuccessCount    int     `json:"success_count"`
	LastFailureTime float64 `json:"last_failure_time"`
}
